Ops ms output jinr tape output fix 11156

todor-ivanov opened the pull request:
Fixes #11156 

#### Status
In development

#### Description
With the current PR two major files are introduced  in the `bin/adhock-scripts` area. Two of them to be used for generic MongoDB connection setup based on the configuration file for a specific micro service. The third one is to be used for  altering the MongoDB documents related to the JINR Tape rules redirection only.   


#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
N/A

#### External dependencies / deployment changes
N/A

amaltaro commented:
@todor-ivanov before you invest too much time on this script, let me give you my opinion here. 

I think we should have a script to do this ONE job here. It does not need to look nice nor to be flexible, it only needs to do the job. For the real effort to have something in WMCore, robust and with a production-quality implementation, we should make it directly to the MSOutput service via the RESTful API that we discussed last Monday.

For the record, here is my take on how to address a very similar use case in MSTransferor: https://github.com/dmwm/WMCore/issues/11220

todor-ivanov committed:
Add scripts for MongoDB generic connection && specific document altering.

Forward parameters from .sh to .py script.

Read rules Info from input file.

Implement rule substitution logic.

Avoid uploading document if not updated && Fix printout.

todor-ivanov commented:
Hi @amaltaro 
> I think we should have a script to do this ONE job here

Yes, This is the script to do one job only. It contains the bare minimum I had to setup in order to be able to access the MongoDB from inside the K8 cluster. The script called `monogoInit.py` is just the part of the ad-hock script which is usable for creating a generic MongoDB connection, regardless if it was in the K8 cluster or MongoDB As A Service. The rest is needed for setting up the environment and executing the `MSOutputDocAlter.py` in such a way that I can feed it with the set of datasets and new rules from the command line. This were exactly the difficulties for connecting to MongoDB I was mentioning yesterday.

amaltaro commented:
I might be missing something then. Can't we simply do:
```
from WMCore.Database.MongoDB import MongoDB
msOutDB = MongoDB(**msOutDBConfig).msOutDB
```

to create a MongoDB object, just as it's done by MSOutput? Of course one needs to have the WMCore environment loaded in the PYTHONPATH and have the basic structure for `msOutDBConfig`, but other than that I don't see why we would need wrapper scripts and a bunch of override options.

The point I am trying to make here is that we should not put much effort on *scripts* that will be used only until we get the RESTful APIs implemented. We should also try to keep such scripts under control, otherwise having too much scripts might mean that we are not even able to find the right script when we need it.

todor-ivanov commented:
Hi @amaltaro I am ready here with all I need for this rule substitution to happen. Please take a quick look, mostly on the logic this script is following and please do not get into the quality of the code too much. Currently it is tested and is doing exactly what it is expected to do. Let me quickly summarize the steps it follows:
* parsing the list of new ruleIds that we were provided and fetching: `newRuleId`, `datasetName`, `oldRuleID`
* querying reqmgr for any request it is aware of that has this dataset as  an output: using the following API: `reqmgr2/data/request?outputdataset=...&mask=RequestName&mask=RequestStatus&mask=OutputDataset`
* takes the workflow name from the result and then queries MongoDB for any records related to this workflow
* once  a record found iterates through all entries in the `OutputMap` and finds the one related to the dataset we started with
* prints the values for `currentRuleId`, `oldRuleID` and `newRuleID` and asks if everything is correct before altering the database
* updates the document 
* uploads it to mongoDB

Please let me know if you have any remarks on this logic. I am planning to take actions tomorrow morning CERN time. Thank you in advance!

amaltaro commented:
Your logic sounds good to me, Todor. One might consider two/three small enhancements and/or protections here:
1) AFAIK we do not create MSOutput document for `Resubmission` workflows (ACDC). When you query ReqMgr2 for workflows that produced dataset ABC, you might want to filter out `RequestType=Resubmission`.
2) Ensure that such requests are sitting in `announced` status.
3) maybe, in addition to matching the old rule id, you might want to confirm that that rule was created against T1_RU_JINR_Tape.

The script might go well without any of those though, but I just wanted to highlight it.

todor-ivanov commented:
Thanks @amaltaro !

These comments make a lot of sense indeed. I'll add the `Resubmission` filter only and for the rest I will look/watch out while  iterating through the list of rules. This information will be printed on the screen before I confirm the swapping of the rules.

todor-ivanov commented:
Thanks again @amaltaro I did run the script already in production and have swapped the rules for all `announced` workflows, while skipping all resubmissions. But there were a few that  were  in `rejected-archived` status  and I did not touch the MongoDB documents there... Although as I think now, they should be falling under the same category. maybe I should reiterate only for swapping those too. What do you think?

I am uploading the log file from the rule swapping procedure  here and will link this in the issue's comments as well:
[JINR_Tape_Relocation_Rules.log](https://github.com/dmwm/WMCore/files/9166538/JINR_Tape_Relocation_Rules.log)

amaltaro commented:
Thank you for performing this update, Todor.
For the `rejected-archived` workflows, I would suggest not to update them. If they are already in *-archived, that means their MSOutput logic is completed and the mongo record `TransferStatus` attribute should be done.

We do need to come back to the workflows that were updated though, just to make sure they have advanced to the next status (archived). Can you paste here a plain list of with those workflow names (a file in AFS will do). Thanks

amaltaro commented:
I had a look at the log file above, and if I did the right parsing, there are 84 workflows in total that were updated with this operation.

Where 46 of them have already been archived, here is their names:
```
cmsunified_task_EXO-RunIISummer20UL17GEN-00466__v1_T_220212_212035_9074
haozturk_task_HIG-RunIISummer20UL18wmLHEGEN-01437__v1_T_210816_180415_1558
cmsunified_task_EXO-RunIISummer20UL17wmLHEGEN-00544__v1_T_220223_164823_9554
pdmvserv_task_HIG-RunIISummer20UL17wmLHEGEN-00373__v1_T_210531_121305_2244
haozturk_task_HIG-RunIISummer20UL16wmLHEGENAPV-00675__v1_T_210816_184020_5305
haozturk_task_HIG-RunIISummer20UL18wmLHEGEN-01437__v1_T_210816_180415_1558
cmsunified_task_HIG-RunIISummer20UL18GEN-00020__v1_T_210722_133259_8935
cmsunified_task_HIG-RunIISummer20UL18GEN-00020__v1_T_210722_133259_8935
haozturk_task_HIG-RunIISummer20UL18wmLHEGEN-01450__v1_T_210817_101501_1102
pdmvserv_task_HIG-RunIISummer20UL18wmLHEGEN-00234__v1_T_210630_171329_2728
pdmvserv_task_HIG-RunIISummer20UL16wmLHEGEN-00125__v1_T_210531_130028_5353
pdmvserv_task_HIG-RunIISummer20UL18wmLHEGEN-01781__v1_T_210818_170739_942
pdmvserv_task_SUS-RunIISummer20UL16wmLHEGEN-00038__v1_T_210207_212057_3824
pdmvserv_task_EXO-RunIISummer20UL17wmLHEGEN-00806__v1_T_211016_015513_3968
haozturk_task_HIG-RunIISummer20UL16wmLHEGEN-01093__v1_T_210816_180216_5252
pdmvserv_task_PPD-RunIISummer20UL16GENSIM-00001__v1_T_211106_161819_5944
cmsunified_task_HIG-RunIIFall17wmLHEGS-05882__v1_T_210805_164920_2205
haozturk_task_HIG-RunIISummer20UL18wmLHEGEN-01415__v1_T_210816_185818_8371
pdmvserv_Commissioning2021_Cosmics_ReReco_113X_dataRun3_v2_runsel_210921_161508_7872
pdmvserv_Commissioning2021_Cosmics_ReReco_113X_dataRun3_v2_runsel_210921_161508_7872
haozturk_task_TOP-RunIISummer20UL16wmLHEGEN-00192__v1_T_210817_100156_8623
haozturk_task_HIG-RunIISummer20UL16wmLHEGEN-01093__v1_T_210816_180216_5252
haozturk_task_HIG-RunIISummer20UL18wmLHEGEN-01387__v1_T_210816_184726_2866
pdmvserv_task_HIG-RunIISummer20UL18wmLHEGEN-00242__v1_T_210609_155558_6032
haozturk_task_HIG-RunIISummer20UL18wmLHEGEN-01450__v1_T_210817_101501_1102
pdmvserv_task_HIG-RunIISummer20UL17wmLHEGEN-00373__v1_T_210531_121305_2244
pdmvserv_task_TOP-RunIISummer20UL16MiniAODv2-00006__v1_T_220204_141021_792
haozturk_task_HIG-RunIISummer20UL17wmLHEGEN-01410__v1_T_210816_184221_3153
pdmvserv_task_EXO-RunIISummer20UL17wmLHEGEN-00806__v1_T_211016_015513_3968
cmsunified_task_TOP-RunIISummer20UL16MiniAODAPVv2-00246__v1_T_211018_151000_7663
pdmvserv_task_HIG-RunIISummer20UL16wmLHEGEN-00125__v1_T_210531_130028_5353
pdmvserv_task_HIG-RunIISummer20UL18wmLHEGEN-00234__v1_T_210630_171329_2728
haozturk_task_HIG-RunIISummer20UL16wmLHEGENAPV-00675__v1_T_210816_184020_5305
haozturk_task_HIG-RunIISummer20UL17wmLHEGEN-01410__v1_T_210816_184221_3153
cmsunified_task_EXO-RunIISummer20UL16MiniAODv2-00268__v1_T_210813_190714_3277
haozturk_task_HIG-RunIISummer20UL16wmLHEGENAPV-00667__v1_T_210817_100101_5490
pdmvserv_task_HIG-RunIISummer20UL18MiniAODv2-00943__v1_T_220130_135508_2769
pdmvserv_task_SUS-RunIISummer20UL16wmLHEGEN-00038__v1_T_210207_212057_3824
haozturk_task_HIG-RunIISummer20UL16wmLHEGENAPV-00010__v1_T_210817_095913_9960
cmsunified_task_SMP-RunIISummer20UL17MiniAODv2-00054__v1_T_210810_132624_292
pdmvserv_task_HIG-RunIISummer20UL18wmLHEGEN-00242__v1_T_210609_155558_6032
haozturk_task_HIG-RunIISummer20UL16wmLHEGENAPV-00010__v1_T_210817_095913_9960
cmsunified_task_HIG-RunIISummer20UL18MiniAODv2-00574__v1_T_211021_195652_3785
cmsunified_task_BPH-RunIISummer20UL17MiniAODv2-00137__v1_T_210817_114308_3567
pdmvserv_task_HIG-RunIISummer20UL18wmLHEGEN-01781__v1_T_210818_170739_942
haozturk_task_HIG-RunIISummer20UL18wmLHEGEN-01387__v1_T_210816_184726_2866
``` 

and these 38 workflows are still sitting in `announced` status:
```
cmsunified_task_HIG-RunIISummer20UL18GEN-00027__v1_T_210722_124731_5259
haozturk_task_HIG-RunIISummer20UL17wmLHEGEN-01430__v1_T_210816_182210_6909
pdmvserv_task_HIG-RunIISummer20UL17wmLHEGEN-00498__v1_T_210609_155501_4337
haozturk_task_HIG-RunIISummer20UL18wmLHEGEN-01419__v1_T_210817_103428_6241
cmsunified_task_TOP-RunIISummer20UL17MiniAODv2-00121__v1_T_211103_195548_7670
pdmvserv_task_B2G-RunIISummer20UL16wmLHEGENAPV-01406__v1_T_210609_170440_4862
cmsunified_task_HIG-RunIISummer20UL16MiniAODAPVv2-00384__v1_T_211102_155923_6993
haozturk_task_TOP-RunIISummer20UL18wmLHEGEN-00150__v1_T_210816_185306_9789
pdmvserv_task_HIG-RunIISummer20UL17wmLHEGEN-02782__v1_T_210930_105135_1909
cmsunified_task_HIG-RunIISummer20UL18GEN-00027__v1_T_210722_124731_5259
pdmvserv_task_HIG-RunIISummer20UL17wmLHEGEN-00371__v1_T_210609_153047_1600
cmsunified_task_HIG-RunIISummer20UL18MiniAODv2-00177__v1_T_210823_122150_8862
cmsunified_task_HIG-RunIISummer20UL18GEN-00048__v1_T_210707_181451_2121
haozturk_task_HIG-RunIISummer20UL17wmLHEGEN-01430__v1_T_210816_182210_6909
pdmvserv_task_TOP-RunIISummer20UL18wmLHEGEN-00121__v1_T_210628_184548_7844
cmsunified_task_HIG-RunIISummer20UL16MiniAODAPVv2-02245__v1_T_220207_121826_3141
haozturk_task_HIG-RunIISummer20UL16wmLHEGENAPV-00113__v1_T_210817_100005_2405
pdmvserv_task_B2G-RunIISummer20UL18wmLHEGEN-01096__v1_T_210612_200554_7463
pdmvserv_task_HIG-RunIISummer20UL17wmLHEGEN-00371__v1_T_210609_153047_1600
cmsunified_task_EXO-RunIISummer20UL18MiniAODv2-00418__v1_T_210813_181439_66
pdmvserv_task_PPD-Run3Summer21GS-00004__v1_T_211006_112240_4758
pdmvserv_task_HIG-RunIISummer20UL17wmLHEGEN-00498__v1_T_210609_155501_4337
haozturk_task_TOP-RunIISummer20UL18wmLHEGEN-00150__v1_T_210816_185306_9789
haozturk_task_SUS-RunIISummer20UL17wmLHEGEN-00019__v1_T_210721_105714_6070
pdmvserv_task_SUS-RunIISummer20UL16wmLHEGEN-00030__v1_T_210207_212023_6412
pdmvserv_task_TOP-RunIISummer20UL18wmLHEGEN-00121__v1_T_210628_184548_7844
pdmvserv_task_SUS-RunIISummer20UL16wmLHEGEN-00030__v1_T_210207_212023_6412
haozturk_task_HIG-RunIISummer20UL16wmLHEGENAPV-00749__v1_T_210527_122237_9043
cmsunified_task_HIG-RunIISummer20UL18GEN-00048__v1_T_210707_181451_2121
pdmvserv_task_HIG-RunIISummer20UL17wmLHEGEN-02782__v1_T_210930_105135_1909
pdmvserv_task_B2G-RunIISummer20UL16wmLHEGENAPV-01406__v1_T_210609_170440_4862
cmsunified_task_TOP-RunIISummer20UL18wmLHEGEN-00197__v1_T_210517_135509_8957
haozturk_task_HIG-RunIISummer20UL18wmLHEGEN-01419__v1_T_210817_103428_6241
pdmvserv_task_TOP-RunIISummer20UL18wmLHEGEN-00121__v1_T_210628_184548_7844
cmsunified_task_TOP-RunIISummer20UL18wmLHEGEN-00197__v1_T_210517_135509_8957
haozturk_task_HIG-RunIISummer20UL16wmLHEGENAPV-00749__v1_T_210527_122237_9043
haozturk_task_HIG-RunIISummer20UL16wmLHEGENAPV-00113__v1_T_210817_100005_2405
haozturk_task_SUS-RunIISummer20UL17wmLHEGEN-00019__v1_T_210721_105714_6070
```

Spot check on 
haozturk_task_HIG-RunIISummer20UL16wmLHEGENAPV-00113__v1_T_210817_100005_2405

suggests that one of the Tape rules is still waiting for approval. I am trying to follow this up on slack.

amaltaro commented:
Fernando contacted the site admins and those rules waiting in approval should start to move in the coming days.

Todor, as discussed in the past week, I would prefer not to have these scripts in WMCore, because it adds extra responsibility and maintenance on the team and also because it will likely not be necessary in the near future. Instead, I'd suggest to:
* make a MSOutput REST endpoint to perform write operations
* for now, keep a copy of these scripts in any repository that you might have with tools/scripts that you use or constructed.

amaltaro commented:
And now we are left only with the following workflows still sitting in `announced` status:
```
cmsunified_task_HIG-RunIISummer20UL18GEN-00027__v1_T_210722_124731_5259
haozturk_task_HIG-RunIISummer20UL17wmLHEGEN-01430__v1_T_210816_182210_6909
cmsunified_task_TOP-RunIISummer20UL17MiniAODv2-00121__v1_T_211103_195548_7670
cmsunified_task_HIG-RunIISummer20UL18GEN-00027__v1_T_210722_124731_5259
haozturk_task_HIG-RunIISummer20UL17wmLHEGEN-01430__v1_T_210816_182210_6909
pdmvserv_task_TOP-RunIISummer20UL18wmLHEGEN-00121__v1_T_210628_184548_7844
cmsunified_task_HIG-RunIISummer20UL16MiniAODAPVv2-02245__v1_T_220207_121826_3141
pdmvserv_task_PPD-Run3Summer21GS-00004__v1_T_211006_112240_4758
haozturk_task_SUS-RunIISummer20UL17wmLHEGEN-00019__v1_T_210721_105714_6070
pdmvserv_task_SUS-RunIISummer20UL16wmLHEGEN-00030__v1_T_210207_212023_6412
pdmvserv_task_TOP-RunIISummer20UL18wmLHEGEN-00121__v1_T_210628_184548_7844
pdmvserv_task_SUS-RunIISummer20UL16wmLHEGEN-00030__v1_T_210207_212023_6412
pdmvserv_task_TOP-RunIISummer20UL18wmLHEGEN-00121__v1_T_210628_184548_7844
haozturk_task_SUS-RunIISummer20UL17wmLHEGEN-00019__v1_T_210721_105714_6070
```

@todor-ivanov how about passing these workflows over to the DM team for another check to see if they can move forward? Meanwhile, I'd suggest to close this issue out, I don't think there is anything else expected from our side here. What do you think?

todor-ivanov commented:
Hi @amaltaro , agreed.
Indeed, let me put the list of tape rules in non-final state here, so we can have information in one place and I can point DM team to this issue here.


Here is the set of unsatisfied tape rules with the associated workflow names:  
```
 Tape rule:  a3d0916cb70643fd896f88b6291d5278 in non final state:  WAITING_APPROVAL for workflow:  cmsunified_task_HIG-RunIISummer20UL18GEN-00027__v1_T_210722_124731_5259
 Tape rule:  509a1fd70e384be3983f529529f3f4b4 in non final state:  STUCK for workflow:  haozturk_task_HIG-RunIISummer20UL17wmLHEGEN-01430__v1_T_210816_182210_6909
 Tape rule:  76a0dbc27e0d44dd92dd8b6db034cad5 in non final state:  WAITING_APPROVAL for workflow:  cmsunified_task_TOP-RunIISummer20UL17MiniAODv2-00121__v1_T_211103_195548_7670
 Tape rule:  a3d0916cb70643fd896f88b6291d5278 in non final state:  WAITING_APPROVAL for workflow:  cmsunified_task_HIG-RunIISummer20UL18GEN-00027__v1_T_210722_124731_5259
 Tape rule:  509a1fd70e384be3983f529529f3f4b4 in non final state:  STUCK for workflow:  haozturk_task_HIG-RunIISummer20UL17wmLHEGEN-01430__v1_T_210816_182210_6909
 Tape rule:  8c9a63d1421343549c09bc998827cfcc in non final state:  SUSPENDED for workflow:  pdmvserv_task_TOP-RunIISummer20UL18wmLHEGEN-00121__v1_T_210628_184548_7844
 Tape rule:  8ab8144889ff41b49a047938553abfa9 in non final state:  WAITING_APPROVAL for workflow:  cmsunified_task_HIG-RunIISummer20UL16MiniAODAPVv2-02245__v1_T_220207_121826_3141
 Tape rule:  57d7ccf2ee324bf7a339da631f3b65b9 in non final state:  SUSPENDED for workflow:  pdmvserv_task_PPD-Run3Summer21GS-00004__v1_T_211006_112240_4758
 Tape rule:  68ef12b5c8174a0ba8b2d07afd6daf03 in non final state:  SUSPENDED for workflow:  haozturk_task_SUS-RunIISummer20UL17wmLHEGEN-00019__v1_T_210721_105714_6070
 Tape rule:  b16ccc3e1e434114966527efc7ce495e in non final state:  SUSPENDED for workflow:  pdmvserv_task_SUS-RunIISummer20UL16wmLHEGEN-00030__v1_T_210207_212023_6412
 Tape rule:  8c9a63d1421343549c09bc998827cfcc in non final state:  SUSPENDED for workflow:  pdmvserv_task_TOP-RunIISummer20UL18wmLHEGEN-00121__v1_T_210628_184548_7844
 Tape rule:  b16ccc3e1e434114966527efc7ce495e in non final state:  SUSPENDED for workflow:  pdmvserv_task_SUS-RunIISummer20UL16wmLHEGEN-00030__v1_T_210207_212023_6412
 Tape rule:  8c9a63d1421343549c09bc998827cfcc in non final state:  SUSPENDED for workflow:  pdmvserv_task_TOP-RunIISummer20UL18wmLHEGEN-00121__v1_T_210628_184548_7844
 Tape rule:  68ef12b5c8174a0ba8b2d07afd6daf03 in non final state:  SUSPENDED for workflow:  haozturk_task_SUS-RunIISummer20UL17wmLHEGEN-00019__v1_T_210721_105714_6070
```
